Add Kafka heartbeat interval and session timeout to component reference

denisbchrsk opened the pull request:
Thank you for helping make the Dapr documentation better!

**Please follow this checklist before submitting:**
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://docs.dapr.io/contributing/contributing-overview/#developer-certificate-of-origin-signing-your-work))
- [x] [Read the contribution guide](https://docs.dapr.io/contributing/docs-contrib/contributing-docs/)
- [x] Commands include options for Linux, MacOS, and Windows within codetabs
- [x] New file and folder names are globally unique
- [x] Page references use shortcodes instead of markdown or URL links
- [x] Images use HTML style and have alternative text
- [x] Places where multiple code/command options are given have codetabs

In addition, please fill out the following to help reviewers understand this pull request:

## Description

Added documentation on the addition of session timeout and heartbeat interval in Kafka's component references (binding and pubsub).

Relevant to the PR that allows these parameters to be configurable in dapr/components-contrib#3375.

## Issue reference

<!--Please reference the issue this PR will close: #[issue number]-->

denisbchrsk committed:
docs: add session timeout and heartbeat interval to kafka pubsub reference

Signed-off-by: denisbchrsk <155584191+denisbchrsk@users.noreply.github.com>

denisbchrsk committed:
docs: add session timeout and heartbeat interval to kafka binding reference

Signed-off-by: denisbchrsk <155584191+denisbchrsk@users.noreply.github.com>

msfussell commented on daprdocs/content/en/reference/components-reference/supported-bindings/kafka.md:
```diff
@@ -78,6 +82,8 @@ spec:
 | `saslMechanism` | N | Input/Output | The SASL authentication mechanism you'd like to use. Only required if `authtype` is set to `"password"`. If not provided, defaults to `PLAINTEXT`, which could cause a break for some services, like Amazon Managed Service for Kafka. | `"SHA-512", "SHA-256", "PLAINTEXT"` |
 | `initialOffset`   | N | Input | The initial offset to use if no offset was previously committed. Should be "newest" or "oldest". Defaults to "newest". | `"oldest"` |
 | `maxMessageBytes` | N | Input/Output | The maximum size in bytes allowed for a single Kafka message. Defaults to 1024. | `"2048"` |
+| `heartbeatInterval` | N | The interval between heartbeats to the consumer coordinator. Defaults to 3s. | `5s` |
```
Any guidance that you would consider adding here on when to change this setting?

denisbchrsk replied:
Added some guidance per confluent's documentation (https://docs.confluent.io/platform/current/installation/configuration/consumer-configs.html#heartbeat-interval-ms), do you think that this comment is now also relevant to `sessionTimeout`?

denisbchrsk committed:
docs: add recommendation to how to set heartbeat interval

Signed-off-by: denisbchrsk <155584191+denisbchrsk@users.noreply.github.com>

denisbchrsk committed:
Merge remote-tracking branch 'origin/v1.14' into feature/kafka-heartbeat-params

Signed-off-by: denisbchrsk <155584191+denisbchrsk@users.noreply.github.com>

denisbchrsk committed:
docs: small change to heartbeat interval doc

Signed-off-by: denisbchrsk <155584191+denisbchrsk@users.noreply.github.com>

denisbchrsk commented:
Fixed, also changed base branch to v1.14 after the release of v1.13.
Also, this should probably be reviewed after dapr/components-contrib#3375 is approved and merged.

denisbchrsk commented:
dapr/components-contrib#3375 has been merged, @msfussell can you take a look at this again?

msfussell commented on daprdocs/content/en/reference/components-reference/supported-bindings/kafka.md:
```diff
@@ -78,6 +82,8 @@ spec:
 | `initialOffset`   | N | Input | The initial offset to use if no offset was previously committed. Should be "newest" or "oldest". Defaults to "newest". | `"oldest"` |
 | `maxMessageBytes` | N | Input/Output | The maximum size in bytes allowed for a single Kafka message. Defaults to 1024. | `"2048"` |
+| `heartbeatInterval` | N | Input | The interval between heartbeats to the consumer coordinator. The value should be set to at most a 1/3 of `sessionTimeout`'s value. Defaults to 3s. | `5s` |
+| `sessionTimeout`    | N | Input | The maximum time between heartbeats before the consumer is considered inactive and will timeout. Defaults to 10s. | `30s` |
```
Do you really mean "maximum time between heartbeats" here? Shouldn't this say "Maximum session time before all heartbeats with the heartbeatInterval stop" or similar? I would take some of the working from the Kafka documentation https://docs.confluent.io/platform/current/installation/configuration/consumer-configs.html#session-timeout-ms

denisbchrsk replied:
After considering this for a bit, I do agree that it's not exactly clear, so I changed to the following (taking part of Kafka's documentation):
"The timeout used to detect client failures when using Kafka’s group management facility. If the broker fails to receive any heartbeats from the consumer before the expiration of this session timeout, then the consumer will be removed and initiate a rebalance."

I could also take most of the explanation on the `sessionTimeout` from confluent's documentation entirely instead of trying to change it.

msfussell committed:
Merge branch 'v1.14' into feature/kafka-heartbeat-params

denisbchrsk committed:
docs: update sessionTimeout and take new merged changes into consideration

Signed-off-by: denisbchrsk <155584191+denisbchrsk@users.noreply.github.com>

denisbchrsk committed:
docs: remove redundant doc lines

Signed-off-by: denisbchrsk <155584191+denisbchrsk@users.noreply.github.com>

denisbchrsk commented:
It seems that in this PR #4099 which was merged recently, added the documentation of `sessionTimeout` and `heartbeatInterval` as is, directly from the component's metadata.yaml (without the suggested changes in this PR), so I'm assuming for this PR that I'll continue to try to improve the explanation on those fields.

hhunter-ms commented on daprdocs/content/en/reference/components-reference/supported-bindings/kafka.md:
```diff
@@ -93,8 +97,8 @@ spec:
 | `consumerFetchDefault` | N | Input/Output | The default number of message bytes to fetch from the broker in each request. Default is `"1048576"` bytes. | `"2097152"` |
-| `heartbeatInterval` | N | Input/Output | The interval between heartbeats to the consumer coordinator for the client connection. Default is `"3s"`. | `"5s"` |
-| `sessionTimeout` | N | Input/Output | The maximum time between heartbeats before the consumer is considered inactive and will timeout. Default is `"10s"`. | `"20s"` |
+| `heartbeatInterval` | N | Input | The interval between heartbeats to the consumer coordinator. The value should be set to at most a 1/3 of `sessionTimeout`'s value. Defaults to `"3s"`. | `"5s"` |
```
```suggestion
| `heartbeatInterval` | N | Input | The interval between heartbeats to the consumer coordinator. At most, the value should be set to a 1/3 of the `sessionTimeout` value. Defaults to `"3s"`. | `"5s"` |
```

hhunter-ms commented on daprdocs/content/en/reference/components-reference/supported-bindings/kafka.md:
```diff
@@ -93,8 +97,8 @@ spec:
-| `heartbeatInterval` | N | Input/Output | The interval between heartbeats to the consumer coordinator for the client connection. Default is `"3s"`. | `"5s"` |
-| `sessionTimeout` | N | Input/Output | The maximum time between heartbeats before the consumer is considered inactive and will timeout. Default is `"10s"`. | `"20s"` |
+| `heartbeatInterval` | N | Input | The interval between heartbeats to the consumer coordinator. The value should be set to at most a 1/3 of `sessionTimeout`'s value. Defaults to `"3s"`. | `"5s"` |
+| `sessionTimeout` | N | Input | The timeout used to detect client failures when using Kafka’s group management facility. If the broker fails to receive any heartbeats from the consumer before the expiration of this session timeout, then the consumer will be removed and initiate a rebalance. Defaults to `"10s"`. | `"20s"` |
```
```suggestion
| `sessionTimeout` | N | Input | The timeout used to detect client failures when using Kafka’s group management facility. If the broker fails to receive any heartbeats from the consumer before the expiration of this session timeout, then the consumer is removed and initiates a rebalance. Defaults to `"10s"`. | `"20s"` |
```

hhunter-ms commented on daprdocs/content/en/reference/components-reference/supported-pubsub/setup-apache-kafka.md:
```diff
@@ -106,8 +110,8 @@ spec:
 | consumerFetchDefault | N | The default number of message bytes to fetch from the broker in each request. Default is `"1048576"` bytes. | `"2097152"` |
-| heartbeatInterval | N | The interval between heartbeats to the consumer coordinator for the client connection. Default is `"3s"`. | `"5s"` |
-| sessionTimeout | N | The maximum time between heartbeats before the consumer is considered inactive and will timeout. Default is `"10s"`. | `"20s"` |
+| heartbeatInterval | N | The interval between heartbeats to the consumer coordinator. The value should be set to at most a 1/3 of `sessionTimeout`'s value. Defaults to "3s". | `"5s"` |
```
```suggestion
| heartbeatInterval | N | The interval between heartbeats to the consumer coordinator. At most, the value should be set to a 1/3 of the `sessionTimeout` value. Defaults to "3s". | `"5s"` |
```

hhunter-ms commented on daprdocs/content/en/reference/components-reference/supported-pubsub/setup-apache-kafka.md:
```diff
@@ -106,8 +110,8 @@ spec:
-| heartbeatInterval | N | The interval between heartbeats to the consumer coordinator for the client connection. Default is `"3s"`. | `"5s"` |
-| sessionTimeout | N | The maximum time between heartbeats before the consumer is considered inactive and will timeout. Default is `"10s"`. | `"20s"` |
+| heartbeatInterval | N | The interval between heartbeats to the consumer coordinator. The value should be set to at most a 1/3 of `sessionTimeout`'s value. Defaults to "3s". | `"5s"` |
+| sessionTimeout | N | The timeout used to detect client failures when using Kafka’s group management facility. If the broker fails to receive any heartbeats from the consumer before the expiration of this session timeout, then the consumer will be removed and initiate a rebalance. Defaults to "10s". | `"20s"` |
```
```suggestion
| sessionTimeout | N | The timeout used to detect client failures when using Kafka’s group management facility. If the broker fails to receive any heartbeats from the consumer before the expiration of this session timeout, then the consumer is removed and initiates a rebalance. Defaults to "10s". | `"20s"` |
```

hhunter-ms reviewed (changes requested):
quick review

msfussell commented:
@denisbchrsk - Can you please take a look at this review. Would like to get this merged.

msfussell committed:
Merge branch 'v1.14' into feature/kafka-heartbeat-params

msfussell commented:
@denisbchrsk - Another ping here. Thanks

denisbchrsk committed:
docs: fix heartbeatInterval and sessionTimeout

Signed-off-by: denisbchrsk <155584191+denisbchrsk@users.noreply.github.com>

denisbchrsk commented:
@msfussell Apologies for the delay, I've added the changes requested by the review above.

hhunter-ms committed:
Merge branch 'v1.14' into feature/kafka-heartbeat-params

marcduiker commented:
@holopin-bot @denisbchrsk Thank you!
